### PR TITLE
Upgrade notebook Gaussian Process to V4

### PR DESCRIPTION
**Pull request by symeneses**
It solves #79 
The number of particles `n_particles` was changed to 100 to get the notebook executed within an acceptable time.
Thanks to @cluhmann @fonnesbeck for your help 🔥

Co-authored-by: @anarita93

#DataUmbrellaPyMCSprint

#  Gaussian Process notebook

+ [x] Notebook follows style guide https://docs.pymc.io/en/latest/contributing/jupyter_style.html
+ [x] PR description contains a link to the relevant issue: a tracker one for existing notebooks or a proposal one for new notebooks
+ [x] Check the notebook is not excluded from any pre-commit check: https://github.com/pymc-devs/pymc-examples/blob/main/.pre-commit-config.yaml

### Helpful links
* https://github.com/pymc-devs/pymc-examples/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by symeneses** [comment]
We have 2 observations:

1. In the cell:
```py
y_pred = [

    np.random.multivariate_normal(m, S)
    for m, S in zip(gp_trace.posterior["mu_post"].T, gp_trace.posterior["sigma_post"].T)
]
```
We use `Transpose` to get the right dimensions, but not sure if this was the best way to do it.

2. It seems that the notebooks name is misleading, as the name indicates that the notebook is about GP in general, but that's not the case. The other notebooks in the folder are named `GP-XXX`. Should we change the name?

File: myst_nbs/gaussian_processes/gaussian_process.myst.md [hunk 1]
```diff
@@ -6,48 +6,43 @@ jupytext:
...
-    find_MAP,
-    sample,
-    summary,
-    traceplot,
-)
-from theano import shared
-from theano.tensor.nlinalg import matrix_inverse
 
-print(f"Running on PyMC3 v{pm.__version__}")
+from aesara import shared
```
Comment (twiecki) [inline myst_nbs/gaussian_processes/gaussian_process.myst.md:37]: can probably use `pm.MutableData`

File: myst_nbs/gaussian_processes/gaussian_process.myst.md [hunk 2]
```diff
@@ -6,48 +6,43 @@ jupytext:
...
+from aesara.tensor.nlinalg import matrix_inverse
```
Comment (twiecki) [inline myst_nbs/gaussian_processes/gaussian_process.myst.md:38]: There's also https://docs.pymc.io/en/latest/api/generated/pymc.math.matrix_inverse.html#pymc.math.matrix_inverse

File: myst_nbs/gaussian_processes/gaussian_process.myst.md [hunk 3]
```diff
@@ -6,48 +6,43 @@ jupytext:
...
+:author: Chris Fonnesbeck
+:::
 
 +++
 
-Example of simple GP fit, adapted from Stan's [example-models repository](https://github.com/stan-dev/example-models/blob/master/misc/gaussian-process/gp-fit.stan).
+Example of simple Gaussian Process fit, adapted from Stan's [example-models repository](https://github.com/stan-dev/example-models/blob/master/misc/gaussian-process/gp-fit.stan).
 
-This example builds a Gaussian process from scratch, to illustrate the underlying model. PyMC3 now includes a dedicated GP submodule which is going to be more usable for a wider variety of problems.
+This example builds a Gaussian process from scratch, to illustrate the underlying model. PyMC now includes a dedicated Gaussian Process submodule which is going to be more usable for a wider variety of problems.
```
Comment (OriolAbril) [inline myst_nbs/gaussian_processes/gaussian_process.myst.md:27]: I would reword this paragraph. I find the "now" particularly problematic, but in general think the idea can be tranmitted more clearly with something like:

```
For illustrative and divulgative purposes, this example builds a Gaussian process from scratch. However, PyMC includes
a {mod}`module dedicated to Gaussian Processes <pymc.gp>` which is recommended instead of coding everything from scratch.
```

Also added the `mod` reference to link to pymc.gp api page.

File: myst_nbs/gaussian_processes/gaussian_process.myst.md [hunk 4]
```diff
@@ -135,45 +130,53 @@ with gp_fit:
...
 ```
 
-Sample from the posterior GP
+Sample from the posterior Gaussian Process
 
 ```{code-cell} ipython3
 y_pred = [
-    np.random.multivariate_normal(m, S) for m, S in zip(gp_trace["μ_post"], gp_trace["Σ_post"])
+    np.random.multivariate_normal(m, S)
+    for m, S in zip(gp_trace.posterior["mu_post"].T, gp_trace.posterior["sigma_post"].T)
```
Comment (OriolAbril) [inline myst_nbs/gaussian_processes/gaussian_process.myst.md:163]: This is very frustrating because neither numpy nor scipy support generating random draws from a multivariate normal in a batched way. I do know they are working on this though and was planning on waiting until this was fixed to add the multivariate distributions to https://einstats.python.arviz.org/en/latest/api/stats.html. But it wouldn't be very difficult to add a multivariate normal batched random number generator (using the already available standard normal and cholesky functions in einstats).

Is that something that would help? It would obviate the need for the stack a few lines above and for the loop.

Comment (OriolAbril) [reply myst_nbs/gaussian_processes/gaussian_process.myst.md:163]: Extra note: I also think that given how our only goal is plotting, generating `y_pred` values for all posterior draws is a bit too much. The plot later on is a bit too crowded. I think that overlaying 100-200 draws is already enough. 

If you install the [ArviZ development version](https://python.arviz.org/en/latest/getting_started/Installation.html#development) you will be able to do something like:

```
post = az.extract(idata, num_samples=100, rng=rng)
# here loop or use einstats or something else, the extract function defaults to stacking under the hood too
```

File: myst_nbs/gaussian_processes/gaussian_process.myst.md [hunk 5]
```diff
@@ -135,45 +130,53 @@ with gp_fit:
...
 ```
 
 ```{code-cell} ipython3
 for yp in y_pred:
     plt.plot(np.linspace(-6, 6), yp, "c-", alpha=0.1)
-plt.plot(x, y, "r.")
+plt.plot(x, y, "r.");
 ```
 
+*  Adapted from Stan's [example-models repository](https://github.com/stan-dev/example-models/blob/master/misc/gaussian-process/gp-fit.stan) by Chris Fonnesbeck in 2016
```
Comment (OriolAbril) [inline myst_nbs/gaussian_processes/gaussian_process.myst.md:173]: The `## Authors` section title is missing, same for the watermark section.

**Comment by OriolAbril** [comment]
> It seems that the notebooks name is misleading, as the name indicates that the notebook is about GP in general, but that's not the case. The other notebooks in the folder are named GP-XXX. Should we change the name?

I am not very opinionated on this _right now_, but I will oppose changing the name if it comes up in the future. Currently most of the old links are broken (in the sense that they point to v3 docs, but using the version switcher ends up in a 404 page), so changing the name right now has very little effect. Once the updated `www.pymc.io/projects/examples/` links are extended, changing the name will break many references to the docs, so I don't think it would be a good idea to do anymore.

File: myst_nbs/gaussian_processes/gaussian_process.myst.md (hunk 1 above)
Comment (symeneses) [reply myst_nbs/gaussian_processes/gaussian_process.myst.md]: I realized this was not used in the notebook, so I deleted.

File: myst_nbs/gaussian_processes/gaussian_process.myst.md (hunk 2 above)
Comment (symeneses) [reply myst_nbs/gaussian_processes/gaussian_process.myst.md]: Changed!

File: myst_nbs/gaussian_processes/gaussian_process.myst.md (hunk 3 above)
Comment (symeneses) [reply myst_nbs/gaussian_processes/gaussian_process.myst.md]: Done!

File: myst_nbs/gaussian_processes/gaussian_process.myst.md (hunk 4 above)
Comment (symeneses) [reply myst_nbs/gaussian_processes/gaussian_process.myst.md]: `stack` was removed and the for was removed and now `y_pred` is calculated with a vectorized function to remove the `for.`

File: examples/gaussian_processes/gaussian_process.ipynb [hunk 6]
```diff
@@ -4,18 +4,23 @@
    "cell_type": "markdown",
```
[the remark this answers is not in the document]
Comment (symeneses) [reply examples/gaussian_processes/gaussian_process.ipynb:4]: Changed!

**Comment by cluhmann** [comment]
I am not understanding the Read the Docs error message.  It seems to be complaining about another notebooks.  @OriolAbril ?

File: examples/gaussian_processes/gaussian_process.ipynb (hunk 6 above)
Comment (OriolAbril) [inline examples/gaussian_processes/gaussian_process.ipynb:4]: <p>The readthedocs error is due to the post directive. The date needs to be present and in the format shown in the jupyter style guide: <code>mmm DD, YYY</code>.The actual day doesn't really matter, as long as it is relatively close to the last update date, it can be today: <code>Jul 31, 2022</code>  , but could as well be the day the PR was opened or some day next week.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/404/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (OriolAbril) [inline examples/gaussian_processes/gaussian_process.ipynb:4]: <p>I would annotate the dimensions of <code>mu_post</code> and <code>sigma_post</code>. It could even be something as <code>"pred_id"</code> for mu and <code>("pred_id", "pred_id2")</code> for sigma (or pred_id_bis). Or anything really.</p><p> </p><p>I would also move the definition of xgrid, D_pred and D_off_diag to be outside the model context. They are all numpy arrays "only" which I don't think is very clear with the current code structure.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/404/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (OriolAbril) [inline examples/gaussian_processes/gaussian_process.ipynb:4]: <p>I have added it to [<a href="https://einstats.python.arviz.org/en/latest/" target="_blank">xarray-einstats</a> If you install de development version you'll be able to use it already. I would therefore use:</p><p> </p><pre class="ql-syntax" spellcheck="false">from xarray_einstats.stats import multivariate_normal

post = az.extract(gp_trace, num_samples=200)
y_pred = multivariate_normal.rvs(post["mu_post"], post["sigma_post"], dims=("pred_id", "pred_id2"))
</pre><p> </p><p>Both looping and np.vectorize are prefectly fine, but I find it very annoying to 1) need to do it every single time I want samples from a multivariate normal (doesn't really apply to this notebook, but will apply to reader's lives) and 2) loose all the labels and dimension information. Here I tried to see what the data represents in the original example and wasn't able to find it, hence the <code>pred_id</code> suggestions, but in general I do have information which makes it easier for me to work with named dimensions.</p><p> </p><p>Note: I think you will also need to install arviz development version to use <code>az.extract</code> in addition to einstats development version.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/404/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (OriolAbril) [inline examples/gaussian_processes/gaussian_process.ipynb:4]: <p>Finally, update this cell to be:</p><p> </p><pre class="ql-syntax" spellcheck="false">_, ax = plt.subplots(figsize=())
ax.plot(xgrid, y_pred.transpose(..., "sample"), c-", alpha=0.1)
ax.plot(x, y, "r.");
</pre><p> </p><p>Note: the figsize is so the plot becomes a bit bigger (as it used to be in the old notebook) which I think looks nicer, I don't know the figsize that should be used though. If the current smaller one was preferred it could be simplified to</p><p> </p><pre class="ql-syntax" spellcheck="false">plt.plot(xgrid, y_pred.transpose(..., "sample"), c-", alpha=0.1)
plt.plot(x, y, "r.");
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/404/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (OriolAbril) [inline examples/gaussian_processes/gaussian_process.ipynb:4]: <p>There should be a markdown level 2 title right above the watermark code cell. This will make watermark appear on the table of contents (right sidebar) and makes it possible to check the versions a notebook runs on with a single click.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/404/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

**Comment by OriolAbril** [comment]
> I am not understanding the Read the Docs error message. It seems to be complaining about another notebooks. @OriolAbril ?

Already commented above, but I'll add some extra info here for full context. The line right above the error message makes reference to a different notebook yes, but that is a sphinx warning, not an error. It is something that we should be aware of and try to fix it (it might become an actual error in a future version for example), but it is not breaking the doc build. The error references:

```
File "/home/docs/checkouts/readthedocs.org/user_builds/pymc-examples/envs/404/lib/python3.8/site-packages/ablog/post.py", line 40, in _split
    return {s.strip() for s in a.split(",")}
AttributeError: 'NoneType' object has no attribute 'split'
```

which means that what is breaking the doc build is something related to the post directive, it doesn't give specific information about what exactly in the post directive is breaking everything. That I am inferring/deducing from the bad format of the date in the post directive plus the `split(",")` piece; the expected format has a comma, the year only that is currently present does not. My guess is instead of splitting into two substrings, no splitting can happen and a None appears somewhere the ablog extension expects a string and we get the error.

File: examples/gaussian_processes/gaussian_process.ipynb (hunk 6 above)
Comment (symeneses) [reply examples/gaussian_processes/gaussian_process.ipynb:4]: <p>In that case, should we remove also the comments "Posterior mean" and "Posterior covariance"?</p><p> </p>

Comment (OriolAbril) [reply examples/gaussian_processes/gaussian_process.ipynb:4]: <p>I think the comments are ok</p>

Comment (symeneses) [reply examples/gaussian_processes/gaussian_process.ipynb:4]: Updated

Comment (symeneses) [reply examples/gaussian_processes/gaussian_process.ipynb:4]: Modified! and thanks for fixing the issue @OriolAbril

**Comment by OriolAbril** [comment]
Looks like I did not guess the cause of the readthedocs issue 😕 . Can you try adding `ablog<0.10.27` to https://github.com/pymc-devs/pymc-examples/blob/main/requirements-docs.txt#L7? It looks like they made a new release in a date that could explain the errors (I'll try to look into this later on once I have time and see if the release introduced a bug or if we are doing something wrong)

**Comment by symeneses** [comment]
> Looks like I did not guess the cause of the readthedocs issue 😕 . Can you try adding `ablog<0.10.27` to https://github.com/pymc-devs/pymc-examples/blob/main/requirements-docs.txt#L7? It looks like they made a new release in a date that could explain the errors (I'll try to look into this later on once I have time and see if the release introduced a bug or if we are doing something wrong)

I tested locally with ablog 0.10.26 and it worked, so let's see. I added the related issue in the comment.

**Comment by OriolAbril** [comment]
It looks like some of the updated versions (no idea what though) is breaking the thumbnail gallery: https://pymcio--404.org.readthedocs.build/projects/examples/en/404/gallery.html. [This one](https://readthedocs.org/projects/pymc-examples/builds/17519027/) seems to be the last working [build]() for pymc-examples. For now, can you check the versions used there for sphinx and sphinx gallery and pin them? Then open an issue as a reminder to look into this and fix it.

File: myst_nbs/gaussian_processes/gaussian_process.myst.md [hunk 7]
```diff
@@ -96,84 +90,97 @@ N = len(y)
...
+_, ax = plt.subplots(figsize=(12, 8))
+ax.plot(xgrid, y_pred.transpose(..., "sample"), "c-", alpha=0.1)
+ax.plot(x, y, "r.");
 ```
 
+## Authors
+*  Adapted from Stan's [example-models repository](https://github.com/stan-dev/example-models/blob/master/misc/gaussian-process/gp-fit.stan) by Chris Fonnesbeck in 2016
+* Updated by Ana Rita Santos and Sandra Meneses in July, 2022 ([pymc#404](https://github.com/pymc-devs/pymc/pull/404))
+
 ```{code-cell} ipython3
```
Comment (OriolAbril) [inline myst_nbs/gaussian_processes/gaussian_process.myst.md:180]: There should be a watermark title right above this cell: https://docs.pymc.io/en/latest/contributing/jupyter_style.html#watermark

**Review by OriolAbril — APPROVED** [review]

**Comment by symeneses** [comment]
The error in the gallery is still there 😭 . I updated the issue with some more details.

**Comment by OriolAbril** [comment]
I checked the preview and the gallery now and it seems to be working for me. Does it still happen for you after refreshing multiple times or trying a different browser?

**Comment by symeneses** [comment]
> I checked the preview and the gallery now and it seems to be working for me. Does it still happen for you after refreshing multiple times or trying a different browser?

Niiice, it's working fine now. No need to cry 😉
